Improve isodesmic and add element balance constraints

hwpang opened the pull request:
### Motivation or Problem
Add Mark's isodesmic efforts and add element balance constraints.

### Description of Changes
This PR adds two parts: 

(1) attempting to add @amarkpayne 's isodesmic efforts from this branch: https://github.com/ReactionMechanismGenerator/RMG-Py/tree/isodesmic_improvements. As this branch is 4 years old, it is very difficult to rebase or merge. I manually copied his efforts in the isodesmic.py over.

(2)
When @oscarwumit and I are working on the bond additivity correction paper, @am2145 found that the isodesmic reactions found by the current code generate reactions that are not mass balanced. I implemented a fix by adding element balance constraints.

xiaoruiDong commented on test/arkane/encorr/isodesmicTest.py:
```diff
@@ -154,55 +153,69 @@ def test_initializing_constraint_map(self):
-            np.array([6, 3, 1, 1, 6, 0]),
+            benzene_constraints,
+            np.array([1, 3, 6, 3]),
         )
```
I suggest adding a comment indicating what are these, e.g. `[6-ring, C-H, C-C, C=C]`

hwpang replied:
Thanks, done!

xiaoruiDong commented on test/arkane/encorr/isodesmicTest.py:
```diff
@@ -154,55 +153,69 @@ def test_initializing_constraint_map(self):
+            np.array([1, 3, 6, 3]),
         )
+
+        spcs_consts.all_reference_species = [self.propene]
```
Why this is needed?

hwpang replied:
This is to ensure that `reference_constraints` in the code below will only return one thing and so make the check more explicit
```
        spcs_consts.all_reference_species = [self.propene]
        propene_features = spcs_consts._get_all_constraints(self.propene)
        _, reference_constraints = spcs_consts._enumerate_constraints([benzene_features, propene_features])
```

xiaoruiDong commented on test/arkane/encorr/isodesmicTest.py:
```diff
@@ -278,6 +275,8 @@ def test_creating_error_canceling_schemes(self):
+            "rc2",
+            True,
             True,
             True,
```
probably using kwargs to improve readability.

hwpang replied:
Thanks, done

xiaoruiDong commented on test/arkane/encorr/isodesmicTest.py:
```diff
@@ -213,36 +226,20 @@ def test_calculating_constraints(self):
-                    [6, 6, 0, 0, 0, 0, 6, 3, 0, 3, 0, 1],
+                    [ 0,  0,  1,  6,  0,  1,  0,  0,  0,  3,  6,  0,  0],
+                    [ 0,  0,  3, 10,  0,  0,  0,  0,  0,  4, 10,  0,  0],
+                    [ 0,  1,  3,  6,  0,  3,  0,  0,  0,  6,  6,  0,  0],
```
I suggest adding the results of `[feat.__repr__() for feat in caffeine_features]` as a comment in this section, to indicate what these numbers are?

hwpang replied:
Thanks, done

xiaoruiDong reviewed (changes requested):
@hwpang Good work finalizing the isodesmic reaction functionality. I have a few comments, majorly focused on improving readability of this piece of codes.

I foresee there will be cases some users ask about relevant questions or a developer need to modify something here for whatever reason, but future developers may not have adequate background knowledge or experience.

xiaoruiDong commented:
Another general question is, in this implementation, the features (constraints) of reference species will be regenerated for each target molecule in a thermo calculation workflow, right?

xiaoruiDong commented:
After the discussion with Hao-Wei today, we decided to remove the usage of `pyomo`. I further looked at the implementation with `lpsolve`, and think it is possible to further replace `lpsolve` with `scipy.optimize.milp` see the following:
```
# necessary imports
import numpy as np
from scipy.optimize import LinearConstraint, milp

constraints = LinearConstraint(A=c_matrix.T, lb=targets, up=targets)
result = milp(sum_constraints, integrality=1, constraints=constraints, options={'time_limit': 10})

if result.status == 0:
  solution = result.x
```
This API is added in SciPy 1.9.0, which only supports Python >=3.8. No need to implement this right away in this PR, but worth to leave a note here for future reference.

JacksonBurns commented:
> After the discussion with Hao-Wei today, we decided to remove the usage of `pyomo`. I further looked at the implementation with `lpsolve`, and think it is possible to further replace `lpsolve` with `scipy.optimize.milp` see the following:
> 
> ```
> # necessary imports
> import numpy as np
> from scipy.optimize import LinearConstraint, milp
> 
> # Actual implementation
> constraints = LinearConstraint(A=c_matrix.T, lb=targets, up=targets)
> result = milp(sum_constraints, integrality=1, constraints=constraints, options={'time_limit': 10})
> 
> if result.status == 0:
>   solution = result.x
> ```
> 
> This API is added in SciPy 1.9.0, which only supports Python >=3.8. No need to implement this right away in this PR, but worth to leave a note here for future reference.

Chiming in to say that I strongly support switching this to scipy - if we merge #2553 first we could upgrade to the latest Python and possibly do this now? Thoughts @hwpang @xiaoruiDong ?

oscarwumit commented:
Any updates? We should probably get the fix merged as the paper gets published in case someone wants to use the IDR feature.

hwpang commented:
I haven't gotten around to fix this due to Chemprop v2.0 release. If we can't get this resolve before the paper is published, we can direct the user to use the original branch, rather than main.

xiaoruiDong commented:
> I haven't gotten around to fix this due to Chemprop v2.0 release. If we can't get this resolve before the paper is published, we can direct the user to use the original branch, rather than main.

Maybe addition to pointing the user to the original branch, we can also say it will be included in RMG version > 3.2

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+        low_level_h_rxn = (
+            sum(
+                spec[0].low_level_hf298.value_si * spec[1]
+                for spec in self.species.items()
```
Done

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+                            return False
+                        connections.pop(i)
+
+                    return True
```
Done

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+
+    def __repr__(self):
+        symbols = ["", "-", "=", "#"]
+        return f"{self.atom1}{symbols[self.bond_order]}{self.atom2}"
```
We currently are not kekulizing input species for the user. I believe that isodesmic reaction method requires the molecule to be kekulized. @oscarwumit Please correct me if I'm wrong. We currently rely on the user to know how to use the isodesmic method correctly by inputting a kekulized structure.

Alternatively, we could add a pre-processing step in the `ErrorCancelingSpecies` class to force kekulizing the input molecule. What do you think?

oscarwumit replied:
When you say "kekulizing input species" do you mean changing the benzene bonds to single and double bonds? e.g., Cn1cccc1Nc1ccccn1 to CN1C=CC=C1NC1=NC=CC=C1？

If so, I think you are correct. We should either raise an error or add a pre-processing (probably better) to make sure the correct structure is being used for IDR scheme. Thanks.

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+                    return True
+            return False
+
+        if isinstance(other, GenericConstraint):
```
I don't think we compare a bond constraint to a generic constraint within our code, but if a user were to make such constraint, this makes sure to catch that case. @oscarwumit Can you double check on this?

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+    atom1 = AtomConstraint(label=atom1.symbol)
+    atom2 = AtomConstraint(label=atom2.symbol)
+
+    return BondConstraint(atom1=atom1, atom2=atom2, bond_order=int(bond.order))
```
Thanks for the question. It is not for the aromatic bond. Single and double bonds are saved as float in `Bond` object, 

```
mol.atoms[0].bonds
{<Atom 'C'>: <Bond "1.0">, <Atom 'C'>: <Bond "2.0">, <Atom 'H'>: <Bond "1.0">}
```

and float would cause this part of code to fail:

```
def __repr__(self):
        symbols = ["", "-", "=", "#"]
        return f"{symbols[self.bond_order]}{self.atom}"
```

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
-            species (ErrorCancelingSpecies): Species whose constraints are to be enumerated
+        # Initialize list of empty lists. Be careful to avoid making references to a singular empty list
+        for _ in range(len(full_constraints_list)):
+            enumerated_constraints.append([])
```
Thanks, done!

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+                new_constraints_list.append(new_spcs_constraints)
+
+            # Finally, update the new list
+            full_constraints_list = new_constraints_list
```
I'm putting the while loop below for easy reference. As the majority of this code was written by @amarkpayne, I'm interpreting the code based on my understanding of the method. I think all this function is doing is to (1) limit the scope of the reference species if `self.limit_scope`, i.e., keeping reference species whose constraints are within the scope of the target species, and (2) return the counts of unique constraints in each species. I have pushed a version of this code to improve the readability and tested locally to make sure that they should be equivalent.

```
        # Begin enumerating constraints
        while True:
            if (
                len(full_constraints_list[0]) == 0
            ):  # We have exhausted target constraints
                if self.limit_scope:  # No need to enumerate any further
                    break  # Out of the while loop

            # Find a constraint to search for next
            for spcs_constraints in full_constraints_list:
                if len(spcs_constraints) != 0:
                    constraint = spcs_constraints[0]
                    break  # Out of the for loop

            else:
                break  # No more constraints, so break out of the while loop

            # enumerate for each species
            new_constraints_list = []
            for i, spcs_constraints in enumerate(full_constraints_list):
                new_spcs_constraints = [c for c in spcs_constraints if c != constraint]
                matching_constraints = len(spcs_constraints) - len(new_spcs_constraints)
                enumerated_constraints[i].append(matching_constraints)
                new_constraints_list.append(new_spcs_constraints)

            # Finally, update the new list
            full_constraints_list = new_constraints_list
```

My version:

```
    def _enumerate_constraints(self, full_constraints_list):
        """
        Return the target constraint counts and the reference constraint counts.
        """

        target_constraints = full_constraints_list[0]
        reference_constraintss = full_constraints_list[1:]

        # Enumerate through the constraints of reference species and keep only those that are present in the target
        enumerated_reference_constraintss = []

        if self.limit_scope:

            for i, ref_spc in enumerate(self.all_reference_species):

                if not all(constraint in target_constraints for constraint in reference_constraintss[i]):
                    continue

                self.reference_species.append(ref_spc)
                enumerated_reference_constraintss.append(reference_constraintss[i])

        else:
            self.reference_species = self.all_reference_species
            enumerated_reference_constraintss = reference_constraintss
        
        # Get a list of the unique constraints sorted by their string representation
        if self.limit_scope:

            # The scope of constraints to consider is the target constraints
            unique_constraints = self._get_unique_constraints(target_constraints)
            unique_constraints.sort(key=lambda x: x.__repr__())

        else:
            all_constraints = target_constraints + [constraint for constraints in enumerated_reference_constraintss for constraint in constraints]
            unique_constraints = self._get_unique_constraints(all_constraints)
            unique_constraints.sort(key=lambda x: x.__repr__())
        
        # Get the counts of each unique constraint in the target and reference constraints
        target_constraint_counts = [target_constraints.count(c) for c in unique_constraints]
        reference_constraint_counts = []

        for i, ref_constraints in enumerate(enumerated_reference_constraintss):
            reference_constraint_counts.append([ref_constraints.count(c) for c in unique_constraints])

        return target_constraint_counts, reference_constraint_counts
```

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+            new_constraints = []
+
+            if charge < 0:
+                allowable_charges = list(range(charge, 0))
```
The current code limits that if the target species is negatively charged, the reference species has to be negatively charged, but if the target species is positively charged, both 0 charge and positively charged species are allowed. I agree that this asymmetry is a bit weird and inconsistent.

I think we should make it such that negatively charged target species must use negatively charged reference species, positively to positively, and neutral to neutral. I will modify the code accordingly.

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -155,89 +176,359 @@ def calculate_target_thermo(self):
+        all_elements = set()
+        for spc in self.reference_species:
+            all_elements.update(spc.molecule.get_element_count().keys())
+        all_elements.update(self.target.molecule.get_element_count().keys())
```
Thanks, done!

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -302,95 +633,133 @@ def _find_error_canceling_reaction(self, reference_subset, milp_software=None):
 
         for solver in milp_software:
-            if solver == 'pyomo':
+            if solver == "pyomo":
```
@xiaoruiDong I'm working on switching over to use scipy.optimize.milp, but it seems that I kept getting `scipy                     1.7.3` when I install the RMG environment. Did you have to sepcify the scipy version in your PR https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2623?

hwpang commented:
The scipy.optimize.milp function is only available with scipy >= 1.9, which is available only with python >= 3.8. This is not available with the current environment.yml. @JacksonBurns I think it would be easier to develop this after we are able to bump the python version to >= 3.8. What is the progress on this?

hwpang commented on environment.yml:
```diff
@@ -65,6 +65,7 @@ dependencies:
   - networkx
   - pytest
   - pytest-cov
+  - pyutilib
```
Thanks, I removed this as we decide to switch to `scipy`

hwpang commented on arkane/encorr/isodesmic.py:
```diff
@@ -441,9 +817,18 @@ def multiple_error_canceling_reaction_search(self, n_reactions_max=20, milp_soft
+                        new_queue.append(s)
+
+                subset_queue = new_queue
+
```
Thanks for the comment. My understanding is that it is to try to only use subsets of reference species that would not allow for already found reactions. I have re-written this function to improve readability.

Original:

```
        subset_queue = [np.arange(0, len(self.reference_species))]
        reaction_list = []

        while (len(subset_queue) != 0) and (len(reaction_list) < n_reactions_max):
            subset = subset_queue.pop()
            if len(subset) == 0:
                continue
            reaction, subset_indices = self._find_error_canceling_reaction(
                subset
            )
            if reaction is None:
                continue
            else:
                reaction_list.append(reaction)

                for index in subset_indices:
                    subset_queue.append(np.delete(subset, index))

                # Clean up the queue to remove subsets that would allow for already found reactions
                new_queue = []
                reaction_indices = [subset[i] for i in subset_indices]
                for s in subset_queue:
                    if not all([i in s for i in reaction_indices]):
                        new_queue.append(s)

                subset_queue = new_queue
```

My re-write:

```
        subset_queue = [np.arange(0, len(self.reference_species))] # indices of subsets of reference species
        reaction_list = []

        while subset_queue and len(reaction_list) < n_reactions_max:
            current_subset = subset_queue.pop()
            if len(current_subset) == 0:
                continue
            
            reaction, used_subset_indices = self._find_error_canceling_reaction(current_subset) # used_subset_indices are the indices of the current subset that were used in the reaction
            
            if reaction is None:
                continue
            
            reaction_list.append(reaction)

            # Generate new subsets by removing each of the used_subset_indices from the current subset
            for index in used_subset_indices:
                new_subset = np.delete(current_subset, index)
                subset_queue.append(new_subset)

            # Remove subsets that include all of the already used reference species
            used_ref_spc_indices = {current_subset[i] for i in used_subset_indices}
            subset_queue = [
                s for s in subset_queue
                if not used_ref_spc_indices.issubset(s)
            ]
```

hwpang commented:
> Any updates? We should probably get the fix merged as the paper gets published in case someone wants to use the IDR feature.

@oscarwumit Thanks for checking in. I have addressed most of the comments given by @xiaoruiDong. 

However, I am not able to switch over to use `scipy.optimize.milp`, because it requires python >= 3.8 and currently we can only get python = 3.7. It is blocked by https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2635.

@oscarwumit Let me know what do you think we should do. I have tried to hack my way in an attempt to obtain a rmg env with python >= 3.8 so that I can develop for this PR, but I had no success.

JacksonBurns commented:
I'll prioritize getting #2631 and #2641 done so that we can merge #2635 and then this PR

oscarwumit commented:
Thanks for your efforts on this. Let's merge after Python upgrading.

oscarwumit reviewed (changes requested):
I think you are correct about the kekulizing part. As you suggested, please either raise an error or add a pre-processing step.

JacksonBurns commented:
Development moved to #2695 which adds support for the required version of Python needed for the required SciPy version, as well as removes LPsolve
